stack-graphs/src/cycles.rs

```diff
@@ -127,3 +135,150 @@ where
         true
     }
 }
+
+// ----------------------------------------------------------------------------
+// Cycle detector
+
+pub trait Appendable {
+    type Ctx;
+
+    fn append_to(
+        &self,
+        graph: &StackGraph,
+        partials: &mut PartialPaths,
+        ctx: &mut Self::Ctx,
+        path: &mut PartialPath,
+    ) -> Result<(), PathResolutionError>;
+    fn start_node(&self, ctx: &mut Self::Ctx) -> Handle<Node>;
+    fn end_node(&self, ctx: &mut Self::Ctx) -> Handle<Node>;
+}
+
+#[derive(Clone)]
+pub struct AppendingCycleDetector<A> {
+    appendages: List<A>,
+}
+
+pub type Appendages<A> = ListArena<A>;
+
+impl<A: Appendable + Clone> AppendingCycleDetector<A> {
+    pub fn new() -> Self {
+        Self {
+            appendages: List::empty(),
+        }
+    }
+
+    pub fn from(appendages: &mut Appendages<A>, appendage: A) -> Self {
+        let mut result = Self::new();
+        result.appendages.push_front(appendages, appendage);
+        result
+    }
+
+    pub fn append(
+        &mut self,
+        graph: &StackGraph,
+        partials: &mut PartialPaths,
+        ctx: &mut A::Ctx,
```

Comment:
Not in scope for this PR: I'm thinking at some point it would make sense to collapse all of these arena-holders into a single type.  I had originally created them separately thinking that finer-grained types would help avoid some (amorphous, not specific) potential borrow checker issues, and would allow you to throw away e.g. a `PartialPaths` arena while retaining a `StackGraph` for later operations.  But that's not how we're using it, and it would simplify large parts of the crate to only have to pass in a single arenas-holder.